### DISCUSSION TITLE
Expose GDAL VRT connection string parameters in various image endpoints

### DISCUSSION THREAD

**Discussion by TomAugspurger** [developmentseed/titiler#960] [General]
I'd like to visualize a band out of a large GRIB2 file, or perhaps a subdatset out of a NetCDF file. I can do this efficiently if I can build a [GDAL VRT connection string](https://gdal.org/drivers/raster/vrt.html#vrt-connection-string) to hand off to rasterio:

`gdalinfo "vrt:///vsicurl/https://noaahrrr.blob.core.windows.net/hrrr/hrrr.20240620/conus/hrrr.t00z.wrfsfcf00.grib2?bands=123"`

(without the `?bands=123` then rasterio / GDAL needs to read all the messages, which is slow)

I'm using titiler-pgstac, and can override [`PgSTACReader._get_asset_info`](https://github.com/stac-utils/titiler-pgstac/blob/6d84641beda8ab28a79a1864d7700b30926c141a/titiler/pgstac/reader.py#L69) to build an AssetInfo with a URL that's actually a VRT connection string. From there things just work!

I'd like to propose two new parameters to TiTiler's image endpoints, and am curious if it makes sense to "standardize" these in TiTiler itself.

| parameter | type | description |
| --- | --- | --- |
| subdataset_name | string | The name of a subdataset in the source asset to load. |
| subdataset_bands | list[int] | The band numbers (between 1 and N) of the source asset to load. |

For reasons that aren't entirely clear to me, the various "subdatasets" inside of the HRRR grib files are considered bands of a single dataset, while for a NetCDF file like this GOES asset, they're considered subdatasets:

```
❯ gdalinfo "/vsicurl/https://goeseuwest.blob.core.windows.net/noaa-goes18/ABI-L2-CMIPM/2024/205/07/OR_ABI-L2-CMIPM1-M6C01_G18_s20242050728250_e20242050728307_c20242050728373.nc"
Driver: netCDF/Network Common Data Format
Files: /vsicurl/https://goeseuwest.blob.core.windows.net/noaa-goes18/ABI-L2-CMIPM/2024/205/07/OR_ABI-L2-CMIPM1-M6C01_G18_s20242050728250_e20242050728307_c20242050728373.nc
Size is 512, 512
Metadata:
  NC_GLOBAL#cdm_data_type=Image
...
  NC_GLOBAL#title=ABI L2 Cloud and Moisture Imagery
Subdatasets:
  SUBDATASET_1_NAME=NETCDF:"/vsicurl/https://goeseuwest.blob.core.windows.net/noaa-goes18/ABI-L2-CMIPM/2024/205/07/OR_ABI-L2-CMIPM1-M6C01_G18_s20242050728250_e20242050728307_c20242050728373.nc":CMI
  SUBDATASET_1_DESC=[1000x1000] toa_lambertian_equivalent_albedo_multiplied_by_cosine_solar_zenith_angle (16-bit integer)
  SUBDATASET_2_NAME=NETCDF:"/vsicurl/https://goeseuwest.blob.core.windows.net/noaa-goes18/ABI-L2-CMIPM/2024/205/07/OR_ABI-L2-CMIPM1-M6C01_G18_s20242050728250_e20242050728307_c20242050728373.nc":DQF
  SUBDATASET_2_DESC=[1000x1000] status_flag (8-bit integer)
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  512.0)
Upper Right (  512.0,    0.0)
Lower Right (  512.0,  512.0)
Center      (  256.0,  256.0)
```

Perhaps if you had a 3-dimensional array in the NetCDF file then you'd end up with *both* subdatasets and bands?

A couple notes:

1. `subdataset_bands` isn't 100% technically correct. This is the `bands` parameter passed to the VRT query string.  `vrt_bands` would be a bit more accurate, but I prefer the `subdataset` prefix. If we prefer `vrt_bands` then I'd recommend using `vrt_subdataset` or `vrt_subdataset_name` for the other parameter.
2. GDAL does have a `sd` parameter for selecting a subdataset by position. I've chosen not to add it since the docs note
> This mode is for convenience only, please use sd_name to choose a subdataset by name explicitly

**Comment by vincentsarago** [comment]
thanks for starting the discussion @TomAugspurger 

To be honest I'm not quite sure I see how and where to add those parameters, especially if this is to be used in titiler-pgstac. 

If it's to be used in a simple TilerFactory created endpoints (e.g. /cog) then I do think it's pretty simpler for user to create there own `path_dependency` 

```python
def DatasetPathParams(
    url: Annotated[str, Query(description="Dataset URL")],
    sds_name: Annotated[str, Query(description="The name of a subdataset in the source asset to load.")],
    sds_bidx: Annotated[List[int], Query(description="The band numbers (between 1 and N) of the source asset to load."),
) -> str:
    """VRT URL from args""
    ...


app = FastAPI(title="My simple app")

cog = TilerFactory(path_dependency=DatasetPathParams)
app.include_router(cog.router)
```

**Reply by TomAugspurger** [reply]
I might be mistaken, but I think our TiTiler deployment doesn't expose `/cog` route (I think we only support access through a STAC Item).

As you note, getting the parameters through in titiler-pgstac is a challenge. For now, I'm pulling them directly off the `request` object in our subclass of https://github.com/stac-utils/titiler-pgstac/blob/6d84641beda8ab28a79a1864d7700b30926c141a/titiler/pgstac/reader.py#L69.

```python
    vrt_params = {}

    if subdataset_name := request.query_params.get("subdataset_name"):
        vrt_params["sd_name"] = subdataset_name

    if subdataset_bands := request.query_params.getlist("subdataset_bands"):
        vrt_params["bands"] = ",".join([str(band) for band in subdataset_bands])

    if vrt_params:
        params = urllib.parse.urlencode(vrt_params, safe=",")
        asset_url = f"vrt:///vsicurl/{asset_url}?{params}"
``` 

If we go forward with this, I'll like to pass those through properly from the endpoint function to the caller of `_get_asset_info`.

I'm defining my parameters like

```python
@dataclasses.dataclass
class SubdatasetParams(DefaultDependency):
    """Assets, Expression and Asset's band Indexes parameters."""

    subdataset_name: Annotated[
        Optional[str],
        Query(
            title="Subdataset name",
            description="The name of a subdataset within the asset.",
        ),
    ] = None
    subdataset_bands: Annotated[
        Optional[List[int]],
        Query(
            title="Subdataset bands",
            description="The name of a subdataset within the asset.",
        ),
    ] = None
```

adding them to the routes like `/tile` like

```python
    subdataset_params=Depends(self.subdataset_dependency),
```

which makes them available on the request object.

---

I think the main value here would be coordinating on the parameter names used, so that they're standardized across TiTiler deployments. Dunno whether that's worth it or not.

**Reply by vincentsarago** [reply]
how do you pass the request object to the `_get_asset_info` method? 

I don't think it's possible to pass anything to it because it's used in https://github.com/cogeotiff/rio-tiler/blob/f59532afd14abd84dd84d0d0f5dc20947a7181c3/rio_tiler/io/base.py#L455-L543

I understand there is a high level of abstraction between titiler/titiler-pgstac/rio-tiler which makes options passing super hard 😭 

I may have an idea to enable this but it will require some change in rio-tiler. Let me open an issue there.

Just to make it clear, you want an option to be able to `edit` the asset url with additional parameter on user request ?

**Comment by vincentsarago** [comment]
This might appear like pure magic, but here is how you can customize your titiler-pgstac application and add `subdataset_name` and `subdataset_bands` options 😅 

https://gist.github.com/vincentsarago/81f84f1ef349b9ac1a3a383c4424142f


The idea is to create a custom low level `Reader` (accessing the dataset) which takes 2 new arguments:
- subdataset_name
- subdataset_bands

```python
@attr.s
class CustomReader(Reader):

    subdataset_name: Optional[str] = attr.ib(default=None)
    subdataset_bands: Optional[List[int]] = attr.ib(default=None)

    def __attrs_post_init__(self):
        vrt_params = {}
        if self.subdataset_name:
            vrt_params["sd_name"] = self.subdataset_name

        if self.subdataset_bands:
            vrt_params["bands"] = ",".join([str(band) for band in self.subdataset_bands])

        if vrt_params:
            params = urllib.parse.urlencode(vrt_params, safe=",")
            self.input = f"vrt:///vsicurl/{self.input}?{params}"
        super().__attrs_post_init__()
```

those 2 args are passed using the `reader_dependency` from the TilerFactory

- from the backend to the stac reader https://github.com/stac-utils/titiler-pgstac/blob/6d84641beda8ab28a79a1864d7700b30926c141a/titiler/pgstac/factory.py#L227-L232 (in form of `reader_options={'reader_options': {'subdataset_name': 'yo'}}`)
- then from stac reader to the dataset reader https://github.com/stac-utils/titiler-pgstac/blob/6d84641beda8ab28a79a1864d7700b30926c141a/titiler/pgstac/mosaic.py#L344-L348 (in form of `reader_options={'subdataset_name': 'yo'}`

**Reply by TomAugspurger** [reply]
https://github.com/microsoft/planetary-computer-apis/compare/main...user/tom/feature/grib-poc has how I'm testing stuff. Your way is certainly shorter :)

With that, the requests to `tile` include the necessary parameters and we pick them up through the `.request` object when building the AssetInfo.

![image](https://github.com/user-attachments/assets/3f7fcd59-f0c3-451f-9804-5a075285184f)

---

FWIW, if you think these parameters a bit too application-specific to go into TiTiler or TiTiler-pgstac, we can certainly just do it in our application.

**Reply by vincentsarago** [reply]
> FWIW, if you think these parameters a bit too application-specific to go into TiTiler or TiTiler-pgstac, we can certainly just do it in our application.

I think generalizing those options is not so simple! event if we add an option as a dependency, it's mostly certain that you'll still need a custom STACReader with a custom `_get_asset_info` method. The main issue is the number of levels there is when working with PgSTAC mosaic (Mosaic Backend -> STAC Reader -> Dataset Reader) and how to pass options to each ones.

If we where adding default options to the TiTiler TileFactory `path_dependency` this will look like:

```python
def SdsParams(
    url: Annotated[
        str,
        Query(
            ...,
            description="Dataset url",

        ),
    ],
    subdataset_name: Annotated[
        Optional[str],
        Query(
            title="Subdataset name",
            description="The name of a subdataset within the asset.",
        ),
    ] = None,
    subdataset_bands: Annotated[
        Optional[List[int]],
        Query(
            title="Subdataset bands",
            description="The band index of a subdataset within the asset.",
        ),
    ] = None,
):
    vrt_params = {}
    if subdataset_name:
        vrt_params["sd_name"] = subdataset_name

    if subdataset_bands:
        vrt_params["bands"] = ",".join([str(band) for band in subdataset_bands])

    if vrt_params:
        params = urllib.parse.urlencode(vrt_params, safe=",")
        url = f"vrt:///vsicurl/{url}?{params}"

    return url
``` 

which is pretty simple for users to write and add as their own dependencies.

For PgSTAC Mosaic you either need to update the asset url within the `_get_asset_info` method (as you did) or in the low level reader (as I did). Both solution work, but as you noted my solution is a bit shorter and avoid touching the factories.

```python
# Custom ReaderParameter which is used to pass some options to the low-level `Reader`
# used in the PgSTAC Mosaic Backend
# PgSTACMosaicBackend -> STACReader -> Reader
@dataclass(init=False)
class ReaderParams(DefaultDependency):
    """reader parameters."""

    reader_options: Dict = field(init=False)

    def __init__(
        self,
        subdataset_name: Annotated[
            Optional[str],
            Query(
                title="Subdataset name",
                description="The name of a subdataset within the asset.",
            ),
        ] = None,
        subdataset_bands: Annotated[
            Optional[List[int]],
            Query(
                title="Subdataset bands",
                description="The band index of a subdataset within the asset.",
            ),
        ] = None
    ):
        """Initialize ReaderParams """
        params = {}
        if subdataset_name:
            params["subdataset_name"] = subdataset_name

        if subdataset_bands:
            params["subdataset_bands"] = subdataset_bands

        self.reader_options = params

# Custom Low-Level reader which will `edit` the dataset url to use
# the VRT form if parameters are provided
# https://gdal.org/drivers/raster/vrt.html#vrt-connection-string
@attr.s
class CustomReader(Reader):

    subdataset_name: Optional[str] = attr.ib(default=None)
    subdataset_bands: Optional[List[int]] = attr.ib(default=None)

    def __attrs_post_init__(self):
        vrt_params = {}
        if self.subdataset_name:
            vrt_params["sd_name"] = self.subdataset_name

        if self.subdataset_bands:
            vrt_params["bands"] = ",".join([str(band) for band in self.subdataset_bands])

        if vrt_params:
            params = urllib.parse.urlencode(vrt_params, safe=",")
            self.input = f"vrt:///vsicurl/{self.input}?{params}"

        super().__attrs_post_init__()


# Custom PgSTACReader (rio_tiler.io.MultiBaseReader)
# which use the `CustomReader` as low level reader
# This Reader will be used for the /collections/{collection_id}/items/{item_id} endpoints
@attr.s
class PgSTACReader(reader.PgSTACReader):
    reader: Type[BaseReader] = attr.ib(default=CustomReader)


# CustomSTACReader (based on titiler.pgstac.reader.CustomSTACReader)
# which use the `CustomReader` as low level reader
# This Reader will be used for the /collections/{collection_id} or /searches/{search_id} endpoints
@attr.s
class CustomSTACReader(mosaic.CustomSTACReader):

    reader: Type[BaseReader] = attr.ib(default=CustomReader)


# Custom Titiler PgSTAC mosaic backend
# which use the `CustomSTACReader` as low level reader
@attr.s
class PGSTACBackend(mosaic.PGSTACBackend):

    reader: Type[BaseReader] = attr.ib(init=False, default=CustomSTACReader)
```

**Comment by hrodmn** [comment]
Is there precedent for using an existing parameter like `bidx` in a specific way for a particular media type (e.g. GRIB2) at the `rio-tiler` level? You can render specific layers from a GRIB file with the `bidx` parameter but it is very slow because `rasterio` has to load the entire file before plucking out the specific band that you asked for. I don't know enough about the GRIB format to understand why this is the case. In this case we use the `bidx` parameter and substitute the original href for the `vrt://{href}?bands={bidx}` path.

**Comment by vincentsarago** [comment]
FYI over https://gist.github.com/vincentsarago/43b2bf643f7f878761004c8145a5f959 I've explored another way

TL&DR, the custom STACReader in the gist, accepts assets to be passed in form of `vrt://{your asset name}?{vrt_option}`  and will construct the HREF url dynamically 

note: using this in titiler, means that you'll need to encode the asset string (because of the ://, ? and & strings)

**Comment by ghidalgo3** [comment]
Thanks Vincent, let me make sure I understood what the gist is doing:
1. The example ECMWF STAC item has a GRIB asset with named `data`. Its href does not start with `vrt://`. If you were to make a titiler call with `asset=data` then probably the `GDAL` call wouldn't work. 
1. If a user makes a tile request to titiler and sets query param `asset=vrt://{asset_name}?{options}` _then_ the `AssetInfo` created by the `STACReader` uses the request's VRT options. At that point, basically the requestor is in control of the VRT URI, which I think is equivalent to what we do in Planetary Computer with our render configs.
1. The change you would commit to titiler is detecting if the `asset` contains `vrt://` and then using that query param to build the VRT.

Did I understand that right? I like this solution because it just exposes [the VRT URI](https://gdal.org/en/latest/drivers/raster/vrt.html#vrt-connection-string). If GDAL adds more options in the future, titiler doesn't have to change.

I think this approach would also work for NetCDF and HDF5, if it's not too onerous can you try it out with one of our NetCDF datasets from Planetary Computer?

**Comment by vincentsarago** [comment]
> The example ECMWF STAC item has a GRIB asset with named data. Its href does not start with vrt://. If you were to make a titiler call with asset=data then probably the GDAL call wouldn't work.

yes, in addition to the code shared in the gist, I would need to account that maybe some asset could be named `vrt://something` (because I don't think there a limit on what the asset key can be).

 >  If you were to make a titiler call with asset=data then probably the GDAL call wouldn't work.

depends what is the asset `data`, if it point to a file readable by GDAL or not 

> If a user makes a tile request to titiler and sets query param asset=vrt://{asset_name}?{options} then the AssetInfo created by the STACReader uses the request's VRT options. At that point, basically the requestor is in control of the VRT URI, which I think is equivalent to what we do in Planetary Computer with our render configs.

Yes

> The change you would commit to titiler is detecting if the asset contains vrt:// and then using that query param to build the VRT.

Yes 

> Did I understand that right? I like this solution because it just exposes [the VRT URI](https://gdal.org/en/latest/drivers/raster/vrt.html#vrt-connection-string). If GDAL adds more options in the future, titiler doesn't have to change.

absolutely 👍 

> I think this approach would also work for NetCDF and HDF5, if it's not too onerous can you try it out with one of our NetCDF datasets from Planetary Computer?

sure
